carl/context/selection.py

```diff
@@ -0,0 +1,105 @@
+from abc import abstractmethod
+import numpy as np
+from carl.utils.types import Context
+from typing import Dict, Any, Optional, Tuple, List
+
+
+class AbstractSelector(object):
+    """
+    Base class for context selectors.
+
+    Context is selected when calling `select`, not in `__init__`.
+
+
+    Parameters
+    ----------
+    contexts: Dict[Any, Context]
+        Context set. A `Context` is a Dict[str, Any].
+
+
+    Attributes
+    ----------
+    contexts : Dict[Any, Context]
+        Context set.
+    context_ids : List[int]
+        Integer index for contexts.
+    contexts_keys : List[Any]
+        Keys of contexts dictionary.
+    n_calls : int
+        Number of times `select` has been called.
+    context_id : Optional[int]
+        Context id of current selected context. Is None at first.
+
+    """
+    def __init__(self, contexts: Dict[Any, Context]):
+        self.contexts: Dict[Any, Context] = contexts
+        self.context_ids: List[int] = list(np.arange(len(contexts)))
+        self.contexts_keys: List[Any] = list(contexts.keys())
+        self.n_calls: int = 0
+        self.context_id: Optional[int] = None  # holds index of current context (integer index of context keys)
+
+    @abstractmethod
+    def _select(self) -> Tuple[Context, int]:
+        """
+        Select next context (internal).
+
+        Should be implemented in child class, internal use.
+
+        Returns
+        -------
+        context : Context
+            Selected context.
+        context_id : int
+            Integer id of selected context.
+        """
+        ...
+
+    def select(self) -> Context:
+        """
+        Select next context (API).
+
+        Returns
+        -------
+        context : Context
+            Selected context.
+        """
+        context, context_id = self._select()
+        self.context_id = context_id
+        self.n_calls += 1
+        return context
+
+
+class RandomSelector(AbstractSelector):
+    """
+    Random Context Selector.
+    """
+    def _select(self):
+        # TODO seed?
+        context_id = np.random.choice(self.context_ids)
+        context = self.contexts[self.contexts_keys[context_id]]
+        return context, context_id
+
+
+class RoundRobinSelector(AbstractSelector):
+    """
+    Round robin context selector.
+
+    Iterate through all contexts and then start at the first again.
+    """
+    def _select(self):
+        if self.context_id is None:
+            self.context_id = -1
+        self.context_id = (self.context_id + 1) % len(self.contexts)
+        context = self.contexts[self.contexts_keys[self.context_id]]
+        return context, self.context_id
+
+
+class ManualSelector(AbstractSelector):
```

Comment:
@TheEimer I added a manual selector which gets a function that selects the next context. Depending on what we want to do we can specify the selector function or write a new class. And we need to talk again about the "new context" flag or when we should actually update our environment

Reply:
Thanks. So that function is just a reference, right? That means teacher-based things are covered really well and also relatively automated things (e.g. as soon as reward is > 200, change instances). Really manual selection is a bit harder, do we in that case expect people not to do that or do we want them to manually set the context somehow? I guess for debugging purposes somebody might want to, else it's not super important - we should be clear about it for doc purposes, though.

Reply:
Hm, schedule selector makes it sound like the schedule would be fixed, but that's not the case, right? Maybe FunctionSelector? But that sounds like we're selecting functions xD